Add CreateOrUpdate utility method

calind opened the pull request:
Closes #85.

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -46,3 +50,56 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
+}
+
+// TransformFn is a function which take in a kubernetes object and returns the
+// desired state of that object
```
note whether or not it's allowed/expected to mutate the object.  It can be helpful when refactoring later.

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -46,3 +50,56 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
+	OperationUpdated = "updated"
+)
+
+// CreateOrUpdate creates or updates a kuberenes resource. It takes in a key and
```
typo: "kubernetes"

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -46,3 +50,56 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
+
+// CreateOrUpdate creates or updates a kuberenes resource. It takes in a key and
+// a placeholder for the existing object and returns the modified object
+func CreateOrUpdate(ctx context.Context, c client.Client, key client.ObjectKey, existing runtime.Object, t TransformFn) (runtime.Object, OperationType, error) {
```
our client methods tend not to return objects, and instead operate "in-place" on them.  This should do so as well

calind commented:
Hi @DirectXMan12 I've implemented your feedback and also wrote tests

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -46,3 +50,75 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
+
+// CreateOrUpdate creates or updates a kubernetes resource. It takes in a key and
+// a placeholder for the existing object and returns the operation executed
+func CreateOrUpdate(ctx context.Context, c client.Client, key client.ObjectKey, existing runtime.Object, t TransformFn) (OperationType, error) {
```
is the purpose of `OperationType` just logging?

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -46,3 +50,75 @@ func SetControllerReference(owner, object v1.Object, scheme *runtime.Scheme) err
+		// Create a new zero value object so that the in parameter of
+		// TransformFn is always a "clean" object, with only Name and Namespace
+		// set
+		zero := reflect.New(reflect.TypeOf(existing).Elem()).Interface()
```
this is what `existing` is for.  It should never be nil, so you shouldn't have to construct a new zero object.

calind replied:
The reason for creating a new zero value struct is that you can pass in existing an initialized one, eg. with `.spec` filled. I'm not sure that `client.Get` zeroes the object if it doesn't find one.

What I want to enforce here, is that if this function is going to create a new object, your transform callback always gets an empty object with only Name and Namespace filled in.

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil_test.go:
```diff
@@ -44,10 +48,94 @@ var _ = Describe("Controllerutil", func() {
+			By("returning OperationCreated")
+			Expect(op).Should(BeEquivalentTo(controllerutil.OperationCreated))
+
+			By("returning returning no error")
```
typo: duplicate word

DirectXMan12 commented on pkg/controller/controllerutil/controllerutil_test.go:
```diff
@@ -44,10 +48,94 @@ var _ = Describe("Controllerutil", func() {
+
+			var scale int32 = 2
+
+			Expect(c.Create(context.TODO(), depl)).Should(Succeed())
```
style is general `Expect(X).To(Y)` (`Should` is reserved for the omega character, which we don't use)

calind commented:
@DirectXMan12  what do you think about changing the signature to 
```go
CreateOrUpdate (c client.Client, ctx context.Context, obj runtime.Object, t ...TransformFn)
```

That way you can just pass the desired object and if you need to apply some changes you can define a set of transform functions.

The reason for keeping the transform function is that for example in case of Service another controller fills in the `ClusterIP` and you cannot reset it to null (you will get a validation error).

The `TransformFn` would change the signature to:

```go
type TransformFn func(existing, desired runtime.Object, op OperationType) (runtime.Object, error)
```

calind commented:
Also, do you thing that I should move this to a `client/clientutil` package?

DirectXMan12 commented:
>  Also, do you thing that I should move this to a `client/clientutil` package

I think it's probably fine where it is.

>  what do you think about changing the signature to ...

I'm fine with the multiple transform functions.  I'm not entirely certain as to the purpose of passing the operation type to the transform function, though.  Same with passing existing and desired pointers -- ideally, the transformation should be idempotent, and should take an object, and turn it into the desired state.

I'd expect a signature of

```go
func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, transforms... TransformFunc) error
type TransformFunc func(obj runtime.Object) error
```

but it's possible I'm missing something here.

calind commented:
@DirectXMan12 I think I've made some logic mistakes in the previous commit so I rewrote it in 1431a57. Let me try to better illustrate the use cases for transform functions which are actually reconcile functions (I have renamed the type to better illustrate this to `ReconcileFn`).

Probably 90% of times you would use `CreateOrUpdate`  by just giving it a `desired` object. The role of reconcile functions is to cover some corner cases. 

For example if you pass in a `corev1.Service` as a desired object, when the object is created in apiserver, another controller will set it's `spec.clusterIP`. After is set, the field becomes immutable. A reconcile function solves this by copying `spec.clusterIP` from the existing object into the desired object.

So the `ReconcileFn` signature would become:
```go
type ReconcileFn func(desired, existing runtime.Object) error
```

As for the operation I've dropped it from the `ReconcileFn` and mainly it's purpose is for informing the caller what action have been taken (eg. for logging purposes). If you have any feedback on how this should be passed I'm open for changes :)

DirectXMan12 commented:
aaaah, I see what you're saying with the reconcile func.  That makes sense, we just need to make sure that's blindingly obvious.

calind commented:
@DirectXMan12 after using it in real code, I'm having second thoughts about the current state of `ReconcileFn`:
```go
type ReconcileFn func(desired, existing runtime.Object) error
```
because I allways end up with code like:
```go
existing.DeepCopyInto(desired)
desired.replicas = 2
```
Ideally you want your reconcile function to just do:
```go
existing.replicas = 2
```
And that is because you always want to mutate the existing state towards a desired state.

So the proposal is to change the signatures to:
```go
// CreateOrUpdate a resrouce with the same type, name and namespace as obj 
// and recocile it's state using the passed in ReconcileFn.
// It returns the executed operation and an error.
func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, r ReconcileFn) (OperationType, error) {}

// ReconcileFn is a function which mutates the existing object into it's desired state. 
// If there is no existing object, a zero value object is passed in.
type ReconcileFn func(existing runtime.Object) error
```

Here is an example which always creates/updates a deployment with the scale of 2:
```go
func scale(existing runtime.Object) error {
    var replicas int32 = 2
    out := existing.(*appsv1.Deployment)
    existing.Sepc.Replicas = &replicas
}
CreateOrUpdate(context.TODO(), c, &appsv1.Deployment{Name: "foo", Namespace: "bar"}, scale)
```

DirectXMan12 commented:
So, we just need to be clear on the intended usage in the comments -- the `ReconcileFunc` should actually be doing the work -- it's not just reconciling the differences any more.  In the original proposal,
it seemed like you'd do work on the passed in object before calling `CreateOrUpdate`, and then reconcile the differences if it already exists.

calind commented:
@DirectXMan12, I've pushed the changes. I've also added an example. It is ready for review.

calind commented:
I've rebased the changes and also found an issue when the `ReconcileFn` changes the name/namespace of the object.

@DirectXMan12 PTAL

DirectXMan12 commented:
hey, sorry, busy week.  let me try and take a look today.

calind commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -97,3 +101,74 @@ func referSameObject(a, b v1.OwnerReference) bool {
+// It returns the executed operation and an error.
+func CreateOrUpdate(ctx context.Context, c client.Client, obj runtime.Object, r ReconcileFn) (OperationType, error) {
+	// op is the operation we are going to attempt
+	var op OperationType = OperationNoop
```
I think needs to be a way to surface the operation. For example, I'm using it to emit events based of that, eg. `Successfully created Deployment default/foo`. Is there a better way of doing it?

calind commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -97,3 +101,74 @@ func referSameObject(a, b v1.OwnerReference) bool {
+		Name:      mo.GetName(),
+		Namespace: mo.GetNamespace(),
+	}
+	existing := reflect.New(reflect.TypeOf(obj).Elem()).Interface().(runtime.Object)
```
update: after more reading I see that `client.Create` and `client.Update` mutate the `obj`. I'll update the PR accordingly.

calind commented:
@DirectXMan12 PTAL.

I've implemented the feedback regarding `obj` parameter. I've also kept `OperationType` in return as a way to surface the operation. I'm open to suggestions on better ways to do this. 

For example I'm using the `OperationType` to send events, like `Successfully created Deployment default/foo` when creating/updating a dependent deployment, logging being the other obvious use case.

DirectXMan12 commented:
I think this looks good.  Let's have a quick peek from @droot and then we'll merge.

droot commented:
@DirectXMan12 @calind looking at it.

droot commented on pkg/controller/controllerutil/controllerutil.go:
```diff
@@ -97,3 +101,78 @@ func referSameObject(a, b v1.OwnerReference) bool {
+		if reflect.DeepEqual(existing, obj) {
+			return OperationNoop, nil
+		}
+		err = c.Update(ctx, obj)
```
Trying to analyze the semantics of this wrt status sub-resource. If an API implements `status` as sub-resource (which is the recommended approach), then `Update` is updating just the `Spec` part of API (Then the `DeepEqual` should be of the spec's field only ?) @DirectXMan12

calind replied:
@droot  `obj` being a `runtime.Object`, is it possible to get the spec of that? If so what should happen with types not having a `Spec` field?

calind commented:
@droot PTAL

droot commented:
@calind looks good to me. Can you pl. squash the commits ?

calind commented:
@droot done

DirectXMan12 commented:
/lgtm

DirectXMan12 commented:
/approve

blaggacao commented:
@calind Is it correct to assume that I would pass on the _desired_ state object as `runtime.Object` to the `CreateOrUpdate` function?
Like this:
```golang
	// Fetch the OdooCluster instance
	instance := &clusterv1beta1.OdooCluster{}
	// Define the desired DBNamespace object
	objDBNamespace := &clusterv1beta1.DBNamespace{
		ObjectMeta: metav1.ObjectMeta{
			Name:      instance.Name + "-dbnamespace",
			Namespace: instance.Namespace,
		},
		Spec: instance.Spec.DBSpec,
	}

	if err := controllerutil.SetControllerReference(instance, objDBNamespace, r.scheme); err != nil {
		return reconcile.Result{}, err
	}

	_, err := controllerutil.CreateOrUpdate(context.TODO(), r, objDBNamespace, scale)
       // Still don't know how to properly handle _ but doesn't matter, here.

```

calind commented:
The object you pass in is the _receiver_ object, aka the result of `client.Create` or `client.Update`. The desired state must be reconciled in the `MutateFn` (aka. the last parameter) like this:

```go
// Define the object name/namespace
objDBNamespace := &clusterv1beta1.DBNamespace{
	ObjectMeta: metav1.ObjectMeta{
		Name:      instance.Name + "-dbnamespace",
		Namespace: instance.Namespace,
	},
}

result, err := controllerutil.CreateOrUpdate(context.TODO(), r.Client, objDBNamespace, func (existing runtime.Object) error{
	// mutate here the state of existing object to the desired state
	// note that at this point _objDBNamespace_ and _existing_ point to the same struct
	out := existing.(*clusterv1beta1.DBNamespace)

	out.Spec = instance.Spec.DBSpec  // Desired state

	if err := controllerutil.SetControllerReference(instance, out, r.scheme); err != nil {
		return err
	}

	return nil
})

if err != nil {
	return reconcile.Result{}, err
}
// Your objDBNamespace now contains the object from the culster after Create/Update
// see: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/client#Writer
// Result contains the applied operation: Create/Update/NoOp
log.Info("object reconciled", "operation", result)

return reconcile.Result{}, nil
```

blaggacao commented:
Thanks, got it! The inline function idiom is a nice idea to keep logic linear to read!

blaggacao commented:
@calind This commit enabled a beautiful pattern https://github.com/xoe-labs/odoo-operator/blob/2a0248030b53ad82b99c2c229fdd1db3f3ae38aa/pkg/controller/odoocluster/odoocluster_controller.go#L212-L459
I guess I still have to take care of become-immutable fields manually in the mutate function, right?

calind commented:
Cool, glad to hear that! You can verify in the in the mutate function that `out.ObjectMeta.CreationTimestamp.IsZero()` to check that an object is going to be created or updated.
